Fixes custom entries precedence bugs (#11331 part)

herbelin opened the pull request:
**Kind:** bug fix 

These are commits 2, 3 related to #11331 in #11355 (+ change log). 

Fixes / closes #11331
Fixes #9519
Fixes #9517

Added (Feb 10): extended to support that by default not mentioning the level of a custom subentry means referring to the highest level of the corresponding entry (as in camlp5).

Added (Feb 15):
- new short commit that fixes #9521 (8th commit)
- a paragraph in the reference manual to indicate the default behavior for internal subentries
- update of change log

- [X] Added / updated test-suite
- [X] Entry added in the changelog
- [X] Reference manual updated

ppedrot commented on vernac/egramcoq.ml:
```diff
@@ -231,7 +233,7 @@ let adjust_level assoc from = let open Gramlib.Gramext in function
     end
   (* None means NEXT *)
-  | (NextLevel,_) -> Some None
+  | (NextLevel,_) -> (* invariant custom=custom' expected *) Some None
```
Why not an `assert` then?

herbelin replied:
Indeed, the cost is anecdotic, why not. I made the change.

herbelin commented:
The failure of bedrock2 (and fiat-crypto, file `bedrock2/NotationsCustomEntry.v`) shows an impact of a needed change of semantics about which there is actually some design choice room. Before this PR, in
```
Declare Custom Entry bedrock_expr.    
Notation "bedrock_expr:( e )"   := e   (e custom bedrock_expr,  
                                        format "'bedrock_expr:(' e ')'")
```
the level of `e` was set by default to 200 (like it is done in the `constr` entry). Now, it has to be given explicitly (and thus bedrock2 fails). An alternative would be that we take as default the maximal level, whatever it is, since the maximal level can dynamically change by adding new rules at a level higher than those already used in the entry.

Which solution is best: a statically enforced level given explicitly by the user, or the highest level, knowing that it may change when new rules are added?

ppedrot commented:
> Which solution is best: a statically enforced level given explicitly by the user, or the highest level, knowing that it may change when new rules are added?

I'd go for the explicit level, this kind of stuff is always a pain to control otherwise.

JasonGross commented:
That is a question for @andres-erbsen or @samuelgruetter

andres-erbsen commented:
Sounds good. I am quite happy to see this getting fixed.

I think if there was a `level *` (or a default behavior) that meant a level above all all already-used and future to-be-used levels, I would use it. But level accounting isn't the bottleneck for notation usage right now; fixing bugs like this one in any way is just more important (imho).

herbelin commented:
@andres-erbsen: thanks for your answer and sorry for not having seen #9517-#9520 until today (I was busy with other works last winter and I stopped for some while being able to manage the dozens of github mails I was receiving up to then; I finally resolved the issue by stopping reading the mails not explicitly notified to me; this more selective approach seems to work better up to now). I will answer to #9517-#9520 directly there.

Providing a "at highest level" would be feasible at little cost.

ppedrot commented:
The requirement for the explicit level makes this not backportable to 8.11, I believe. Is that a problem?

herbelin commented:
> The requirement for the explicit level makes this not backportable to 8.11, I believe. Is that a problem?

An alternative is to backtrack on the explicit level. What provides the more also provides the less.

andres-erbsen commented:
@herbelin Thanks for working on this. With the current fixes in the pipeline (and optimistically assuming no more issues pop up once we get unblocked), I think we can make bedrock2 user experience as smooth as stand-alone language.

herbelin commented:
@ppedrot:
> The requirement for the explicit level makes this not backportable to 8.11, I believe.

I tried to remove the explicit level requirement, but this requires a little change in the API (type `Extend.production_level`), which would make it not backportable again??

A third alternative would be to temporarily use a little hack (using, say `-1` as a code for the highest level), which may not be worth than the current situation...

What do you think? Fixing this in 8.11 seems very important for (at least) bedrock2.

ppedrot commented:
I think that the backportability should not sacrifice good design choices. For instance, I still believe that the level should be explicit, but if we want to backport the patch, then it should allow not to specify a level in 8.11, do so with a deprecation warning in 8.12, and not at all in 8.13.

In any case, we're stepping in the lands of design choices, so probably asking the various users like @Janno could be helpful. @herbelin can you summarize the possible variants?

herbelin commented:
>  can you summarize the possible variants?

Let me state the problem as the question of supporting default entry assignment to notation variables in a rule. For instance, in:
```
Declare Custom Entry expr.
Notation "'if' e 'then' t 'else' u" := (e,t,u)
   (in custom expr at level 10,
    e constr).
```
is it reasonable that `t` and `u` are automatically inferred by default to be in `custom expr`, with `t` at the first level (with highest value) in `expr` and `u` (because of a default to no associativity) at the level following 10 in `expr`?

Or do we want to force the user to be explicit as e.g. in:
```
Notation "'if' e 'then' t 'else' u" := (e,t,u)
   (in custom expr at level 10,
    t custom expr at highest level,
    u custom expr at next level,
    e constr).
```
Note that, by comparison, entries in `constr` already have a default: it is level `200` for internal variables (i.e. the highest level statically used in prelude) and it depends on associativity for variables at the border of the expression.

Note that the concept of "highest level" has no syntax yet and a syntax would have to be provided. One suggestion is `t custom expr at level *`. Another suggestion could be `t custom expr at any level`. Or whatever.

Note that [`bedrock2/NotationsCustomEntry.v`](https://github.com/mit-plv/bedrock2/blob/master/bedrock2/src/bedrock2/NotationsCustomEntry.v) already intensively uses implicit declaration of level.

> I think that the backportability should not sacrifice good design choices

Note that I can also make a (slightly hackish) specific 8.11 PR which would change neither the 8.11 notation API (i.e. accepting the first `Notation` above with the explained defaults) nor the 8.11 OCaml API (up to fixing the bugs), leaving the design choice question to 8.12.

andres-erbsen commented:
I now realize `t custom expr at level *` is probably a bad idea, because the next character after `*` would likely be `)`.

herbelin commented:
@andres-erbsen: This PR actually fixes also #9517 and this is the reason for bedrock2 failing. I made a backward-compatible bedrock2 [PR](https://github.com/mit-plv/bedrock2/pull/140) so that CI is here happy.

@ppedrot: My feeling is that the currently-implemented semantics for determining the default entries and levels is after all rather natural and acceptable and does not require extra syntax to refer to the highest level (first part of my previous comment).

ppedrot commented:
OK, let's go for the implicit highest level then. It has the advantage of backwards compatibility.

andres-erbsen commented:
I updated bedrock2 https://github.com/mit-plv/bedrock2/commit/3585b23d66ed5b64523d9e4866a65ec23675ac11

herbelin commented:
@ppedrot:
- I added an extra commit fixing also #9521 (a case missing in `Egramcoq.pure_sublevels'`)
- I updated the doc wrt the default inference policy for levels of subentries
- I updated the change log wrt the extended content of the PR

This seems ready on my side, hoping I did not forget anything.

PS: For 8.11, I'll later do a slightly hackish but shorter fix not modifying the API.

@andres-erbsen: fiat-crypto is failing but I suspect that it is because it does not use the version of bedrock2 which contains mit-plv/bedrock2#141? The line where it fails seems indeed to correspond to an older version.

JasonGross commented:
Indeed, fiat-crypto is using an old version of bedrock2.  I'll try bumping it later tonight.

JasonGross commented:
Oops, did not mean to have my PR close this PR

ppedrot commented:
I am postponing this to 8.12+beta1 then, @herbelin when you have time you should open a dedicated backport PR against 8.11.

ppedrot commented:
Will leave a little bit of additional time before merging for last minute comments from @coq/vernac-maintainers.

herbelin commented:
> you should open a dedicated backport PR against 8.11

The 8.11 PR is in #11613. Question about it: what am I supposed to do for the changelog PR number in the 8.11 and the master PR: to use the same number (either the one of master or 8.11) or specific numbers?

Note: I made a minor last change in both PRs: `constr` at the border of a notation in a custom entry is by default `constr at level 200` (while it was `constr` at the higher level a few minutes before). I don't think that this matters in practice, this is just by convention with what is done elsewhere by default for `constr`.

Zimmi48 commented:
> what am I supposed to do for the changelog PR number in the 8.11 and the master PR: to use the same number (either the one of master or 8.11) or specific numbers?

Please make the two changelog entries identical. Otherwise, this will create merge conflicts when preparing the 8.11.1 release notes. You can link to both PRs in this (common) changelog, or choose to refer to the master PR only.

herbelin commented:
> Please make the two changelog entries identical. Otherwise, this will create merge conflicts when preparing the 8.11.1 release notes. You can link to both PRs in this (common) changelog, or choose to refer to the master PR only.

Ah, thanks a lot for the answer. So, I use the same file name `11530-master+fix11331-custom-entries-precedence.rst` with same content in both PRs, right?

herbelin commented:
> Please make the two changelog entries identical.

Done, then (referring to #11530 in both).

herbelin commented:
@ppedrot: I'm very sorry, I made a mistake in my last push to this PR, pushing inadvertantly 3 experimental commits which were on my branch but not supposed to be pushed. Two of these 3 commits cancel themselves, but the last one (actually a copy of #11561) was there by mistake.

I did not investigate but this might be the reason for #11628. I made #11636 to cancel it. Sorry again.
